File: Kora-Hire/backend/schemas/profiles.py
from pydantic import BaseModel, field_serializer
from uuid import UUID
from typing import List, Optional
from datetime import date

class ProfileSummaryBullet(BaseModel):
    order: int
    detail: str

class ProfileBullets(BaseModel):
    liked_bullets: List[ProfileSummaryBullet]
    disliked_bullets: List[ProfileSummaryBullet]

class ProfileOverallScore(BaseModel):
    profile_score: int
    headline: str

class ProfileCategoryScore(BaseModel):
    category_id: UUID
    category_score: int
    category_score_reason: str

class ProfileRubric(BaseModel):
    category_scores: List[ProfileCategoryScore]

class RoleSummary(BaseModel):
    title: str
    company: str
    tenure: int

class ProfileCareerPath(BaseModel):
    years_of_experience: int
    most_recent_role: RoleSummary
    longest_tenured_role: RoleSummary
    average_tenure: int
    number_promotions: int

class RecentRole(BaseModel):
    id: UUID
    application_id: UUID
    title: str
    organization: str
    recency: int
    end_date: Optional[date] = None
    start_date: date

    @field_serializer("start_date", when_used="json")
    def serialize_start_date(self, value: date) -> str:
        return value.strftime("%B %Y")  # e.g. "January 2024"

    @field_serializer("end_date", when_used="json")
    def serialize_end_date(self, value: Optional[date]) -> str:
        if value is None:
            return "Current"
        return value.strftime("%B %Y")

class Profile(BaseModel):
    profile_id: UUID
    application_id: UUID
    applicant_name: str
    applicant_email: str
    linkedin_url: Optional[str]
    portfolio_url: Optional[str]
    github_url: Optional[str]
    applicant_career: ProfileCareerPath
    profile_score: ProfileOverallScore
    profile_bullets: ProfileBullets
    profile_rubric: ProfileRubric
    recent_roles: Optional[List[RecentRole]]

File: Kora-Hire/frontend/src/types/profiles.ts
// ProfileSummaryBullet.ts
export interface ProfileSummaryBullet {
  order: number;
  detail: string;
}

// ProfileBullets.ts
export interface ProfileBullets {
  liked_bullets: ProfileSummaryBullet[];
  disliked_bullets: ProfileSummaryBullet[];
}

// ProfileOverallScore.ts
export interface ProfileOverallScore {
  profile_score: number;
  headline: string;
}

// ProfileCategoryScore.ts
export interface ProfileCategoryScore {
  category_id: string; // UUID as string
  category?: string;
  category_score: number;
  category_score_reason: string;
}

// ProfileRubric.ts
export interface ProfileRubric {
  category_scores: ProfileCategoryScore[];
}

// RoleSummary.ts
export interface RoleSummary {
  title: string;
  company: string;
  tenure: number;
}

// ProfileCareerPath.ts
export interface ProfileCareerPath {
  years_of_experience: number;
  most_recent_role: RoleSummary;
  longest_tenured_role: RoleSummary;
  average_tenure: number;
  number_promotions: number;
}

// RecentRole.ts
export interface RecentRole {
  id: string;
  application_id: string;
  title: string;
  organization: string;
  recency: number;
  start_date: string;     // e.g. "January 2024"
  end_date: string;       // e.g. "Current" or "November 2023"
}

// Profile.ts
export interface Profile {
  profile_id: string; // UUID
  application_id: string; // UUID
  applicant_name: string;
  applicant_email: string;
  linkedin_url?: string;
  portfolio_url?: string;
  github_url?: string;
  applicant_career: ProfileCareerPath;
  profile_score: ProfileOverallScore;
  profile_bullets: ProfileBullets;
  profile_rubric: ProfileRubric;
  recent_roles?: RecentRole[];
}